### PR TITLE
feat: unified script for deploying kind

### LINKED ISSUE #284: Merge kind scripts
https://github.com/snowdrop/k8s-infra/blob/main/kind/kind-reg-ingress.sh
https://github.com/snowdrop/k8s-infra/blob/main/kind/kind-tls-secured-reg.sh 
https://github.com/snowdrop/k8s-infra/blob/main/kind/kind-reg-kourier-knative.sh

### PR DESCRIPTION
**Pull request by jacobdotcosta**
Support for `podman` as CRI provider.

Closes #284

### CODE REVIEW

**Comment by cmoulliard** [comment]
As you are proposing an approach where content is now packaged into different shell scripts, how will it be possible to `curl & bash` remotely a file ? @jacobdotcosta

**Comment by jacobdotcosta** [comment]
The deployment is only 1 script now, `./kind/deploy-kind.sh`.

**Comment by cmoulliard** [comment]
There is something that I do not understand. When I run the script as such, if a cluster exists then we got an error without a command asking to delete it
```
./kind/deploy-kind.sh 
ERROR: failed to create cluster: node(s) already exist for a cluster with the name "kind"
```
If I delete the cluster manually, then the script will directly create a cluster without asking what I would like to provision: registry, ingress/kourier, etc
```
./kind/deploy-kind.sh     
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.25.3) 🖼 
⠈⠁ Preparing nodes 📦                      
```
@jacobdotcosta

**Comment by jacobdotcosta** [comment]
If no parameters are passed the script deploys the default configuration, which would be an nginx ingress.

```bash
$ ./kind/deploy-kind.sh --help
Usage: 
  ./deploy-kind.sh [parameters,...]

Parameters: 
  --help: This help message

  --cluster-name <name>                 Name of the cluster. Default: kind
  --delete-kind-cluster                 Deletes the Kind cluster prior to creating a new one.
  --ingress [nginx,knative]             Ingress to be deployed. One of nginx,knative. Default: nginx
  --knative-version <version>           KNative version to be used. Default: 1.9.0
  --kubernetes-version <version>        Kubernetes version to be install
                                        Default: latest
  --registry-image-version <version>    Version of the registry container to be used. Default: 2.6.2
  --registry-port <port>                Port to publish the registry. Default: 5000
  --server-ip <ip-address>              IP address to be used. Default: 127.0.0.1
  --verbosity <value>                   Logging verbosity (0..9)
                                        A verbosity setting of 0 logs only critical events.
                                        Default: 0
```
@cmoulliard

**Comment by cmoulliard** [comment]
Here is what we should do with the script
- Check prereq tools are there kind, helm, kubectl (and ideally with aligned versions)
- If a kind cluster exists for default name or another name, then we ask to the user if we will delete them or delye a specific one
- We ask questions about k8s version to be installed, like also if they would like to install:
  - IP address of the VM - default: 127.0.0.1
  - docker registry (= default) or docker secured registry
  - ingress (= default) or kourier
  - knative (not installed by default)
  - etc

**Comment by cmoulliard** [comment]
What do you think about integrating too such scripts/functions ?

- Play allows to demo the script itself (speed can be defined like if we want to have a prompt) like to print or print/execute commands: https://github.com/ch007m/my-stonesoup/blob/main/scripts/play.sh
- Common: where will define the colors, functions to log, etc: https://github.com/ch007m/my-stonesoup/blob/main/scripts/common.sh

WDYT ? @BarDweller @jacobdotcosta

**Comment by jacobdotcosta** [comment]
> Here is what we should do with the script
> 
> * Check if prereq tools are there kind, helm, kubectl (and ideally with aligned versions)
> * If a kind cluster exists for default name or another name, then we ask to the user if we will delete them or delete a specific one

Agree with this. The prereq are being checked prior only to stop the script from executing if a major dependency fails.

> * We ask questions about k8s version to be installed, like also if they would like to install:
>   
>   * IP address of the VM - default: 127.0.0.1
>   * verbosity level (= default is 0)
>   * docker registry (= default) or docker secured registry
>   * ingress (= default) or kourier
>   * knative (not installed by default)
>   * etc

I'm not sure our script should stop to ask these questions. If the problem is having it too automated that the user doesn't know what's being installed, I think the solutions should be to remove some of the defaults and request the user to be added when to the command line.

**Comment by cmoulliard** [comment]
There is an issue @jacobdotcosta 
```
pwd = /Users/cmoullia/code/snowdrop/k8s-infra/kin
$ ./deploy.sh
$ ./deploy.sh: line 78: ./kind-reg.sh: No such file or directory
```

**Comment by jacobdotcosta** [comment]
for now I left the original scripts and implemented `deploy-kind.sh`.

**Comment by cmoulliard** [comment]
The command should be improved as `-h` fails
```
./deploy-kind.sh -h
ERROR: failed to create cluster: node(s) already exist for a cluster with the name "kind"
```

**Comment by jacobdotcosta** [comment]
```
./deploy-kind.sh --help                                                                                                                                    SIGINT   merge-kind-scripts 
Usage: 
  ./deploy.sh [parameters,...]

Parameters: 
  --help: This help message

  --cluster-name <name>                 Name of the cluster. Default: kind
  --delete-kind-cluster                 Deletes the Kind cluster prior to creating a new one.
  --ingress [nginx,knative]             Ingress to be deployed. One of nginx,knative. Default: nginx
  --knative-version <version>           KNative version to be used. Default: 1.9.0
  --kubernetes-version <version>        Kubernetes version to be install
                                        Default: latest
  --registry-image-version <version>    Version of the registry container to be used. Default: 2.6.2
  --registry-port <port>                Port to publish the registry. Default: 5000
  --server-ip <ip-address>              IP address to be used. Default: 127.0.0.1
  --verbosity <value>                   Logging verbosity (0..9)
                                        A verbosity setting of 0 logs only critical events.
                                        Default: 0
```

**Comment by jacobdotcosta** [comment]
I've only implemented the `--` switches
`--help` for help
I can add `-h`

**Comment by cmoulliard** [comment]
I suggest to include such a code & to use it too to log the messages using colors, as title, etc: https://github.com/halkyonio/primaza-poc/blob/main/scripts/common.sh#L8-L59. Improvements are welcome !

WDYT ? @jacobdotcosta

**Comment by jacobdotcosta** [comment]
I like the idea, I'll implement it.

**Comment by cmoulliard** [comment]
This commit 206cbed fails for me 
```
./kind/deploy-kind.sh: line 308: conditional binary operator expected
```

**Comment by jacobdotcosta** [comment]
How did you execute it @cmoulliard . I cannot reproduce this error on the current version. Could you update to the latest changes and test again?

**Comment by cmoulliard** [comment]
> I cannot reproduce this error on the current version. Could you update to the latest changes and test again?

On macos I got also an error here using commit 4136092
```
./kind/deploy-kind.sh
./kind/deploy-kind.sh: line 584: conditional binary operator expected
```

**Comment by jacobdotcosta** [comment]
> !! Changing this line on macos fixes the problem `#!/usr/bin/env bash` @jacobdotcosta

Script updated @cmoulliard

**Comment by jacobdotcosta** [comment]
Implemented a command to be executed...

```bash
ERROR: A command must be provided!!!

Usage: 
	./deploy-kind.sh command [parameters,...]

Available commands: 
	install:				Install the kind cluster
	remove:					Remove the kind cluster
```

Full help.

```bash
./kind/deploy-kind.sh 
ERROR: A command must be provided!!!

Usage: 
	./deploy-kind.sh command [parameters,...]

Available commands: 
	install:				Install the kind cluster
	remove:					Remove the kind cluster

Required parameters: 
	--ingress [nginx,kourier]:		Ingress to be deployed. One of nginx,kourier.

Optional parameters: 
	-h, --help:				This help message

	--cluster-name <name>			Name of the cluster. Default: kind
	--delete-kind-cluster			Deletes the Kind cluster prior to creating a new one. Default: No
	--knative-version <version>		KNative version to be used. Default: 1.9.0
	--kubernetes-version <version>		Kubernetes version to be install. Default: latest
	--registry-image-version <version>	Version of the registry container to be used. Default: 2.6.2
	--registry-password <password>		Registry user password. Default: snowdrop
	--registry-port <port>			Port to publish the registry. Default: 5000
	--registry-user <user>			Registry user. Default: admin
	--secure-registry			Secure the docker registry. Default: No
	--server-ip <ip-address>		IP address to be used. Default: 127.0.0.1
	--use-existing-cluster			Uses existing kind cluster if it already exists. Default: No
	-v, --verbosity <value>			Logging verbosity (0..9). Default: 1
						A verbosity setting of 0 logs only critical events.

**Comment by cmoulliard** [comment]
Can the new script support another option/parameter to install it on podman `KIND_EXPERIMENTAL_PROVIDER=podman kind create cluster` ?

**Comment by jacobdotcosta** [comment]
That's exactly what my latest commit does. By selecting the `--provider` parameters one can change between `docker` and `podman`.

```
--provider <provider>		Container Runtime [docker,podman]. Default: docker
```

**Comment by cmoulliard** [comment]
Will it be possible to use the new script like this `curl -s -L "https://raw.githubusercontent.com/snowdrop/k8s-infra/main/kind/kind-reg-ingress.sh" | bash -s y latest kind 0 ${VM_IP}` ? @jacobdotcosta

File: kind/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,753 @@
...
+    log_message "1" "                                                            |_|    "
+    log_message "1" " Kind installation script"
+    log_message "1" ""
+    log_message "1" "- Deploying a local secured (using htpasswd) docker registry"
+    log_message "1" "- Generating a selfsigned certificate (using openssl) to expose the registry as a HTTP/HTTPS endpoint"
+    log_message "1" "- Setting a docker network between the containers: kind and registry and alias \"registry.local\""
+    log_message "1" "- Allowing to access the repository using as address \"registry.local:${REGISTRY_PORT}\" within a pod, from laptop or when a pod is created"
+    log_message "1" "- Exposing 2 additional NodePort: 30000 and 31000"
+    log_message "1" "- Deploying an ingress controller"
+    log_message "1" "- Copying the generated certificate here: $HOME/local-registry.crt"
```
Comment (cmoulliard) [inline kind/deploy.sh:114]: As this text will depend about what the user will choose, I propose to remove it and instead, according to the parameters selected, that we show pat of the creation/installation what the script will currently do

**Review by cmoulliard — CHANGES_REQUESTED** [review]

File: kind/deploy.sh [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+    log_message "0" "Available commands: "
+    log_message "0" "\tinstall:\t\t\t\tInstall the kind cluster"
+    log_message "0" "\tremove:\t\t\t\t\tRemove the kind cluster"
+    log_message "0" ""
+    log_message "0" "Required parameters: "
+    log_message "0" "\t--ingress [nginx,kourier]:\t\tIngress to be deployed. One of nginx,kourier."
+    log_message "0" ""
+    log_message "0" "Optional parameters: "
+    log_message "0" "\t-h, --help:\t\t\t\tThis help message"
+    log_message "0" ""
```
Comment (cmoulliard) [inline kind/deploy.sh:149]: I dont see a parameter to specify the version to be installed of the k8s cluster as this is the case today. According to the kind tool version installed, they support different k8s versions -see: https://github.com/kubernetes-sigs/kind/releases/tag/v0.18.0

**Review by cmoulliard — CHANGES_REQUESTED** [review]

**Review by cmoulliard — CHANGES_REQUESTED** [review]

File: kind/pre-reqs.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#!/bin/sh
+
+set -o errexit
+
```
[the remark this answers is not in the document]
Comment (jacobdotcosta) [reply kind/pre-reqs.sh:4]: > Will it be possible to use the new script like this `curl -s -L "https://raw.githubusercontent.com/snowdrop/k8s-infra/main/kind/kind-reg-ingress.sh" | bash -s y latest kind 0 ${VM_IP}` ? @jacobdotcosta

Yes.

```bash
curl -s -L "https://raw.githubusercontent.com/snowdrop/k8s-infra/0000c6b460bac45d11e40cf44f1e807a90921e51/kind/deploy.sh" | bash -s install --ingress nginx --cluster-name twm --provider podman --port-map "30000:30000,31000:31000" -v 5
```

File: kind/deploy.sh (hunk 1 above)
Comment (jacobdotcosta) [reply kind/deploy.sh]: This text has been removed.

File: kind/pre-reqs.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacobdotcosta) [reply kind/pre-reqs.sh:4]: I think the `pre-reqs.sh` and `undeploy-kind.sh` scripts can be removed.

**Comment by jacobdotcosta** [comment]
> Will it be possible to use the new script like this `curl -s -L "https://raw.githubusercontent.com/snowdrop/k8s-infra/main/kind/kind-reg-ingress.sh" | bash -s y latest kind 0 ${VM_IP}` ? @jacobdotcosta

Yes.

```bash
curl -s -L "https://raw.githubusercontent.com/snowdrop/k8s-infra/0000c6b460bac45d11e40cf44f1e807a90921e51/kind/deploy.sh" | bash -s install --ingress nginx --cluster-name twm --provider podman --port-map "30000:30000,31000:31000" -v 5
```

File: kind/README.adoc [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+== Introduction
+
+The goal of this project is to simplify our life when we use kubernetes
+on a laptop. It currently supports the following features: 
+
+* Create/delete a K8s cluster using the latest cluster version, or the one you specify, 
+* Launch a local container registry exposed on port `localhost:5000`, 
+* Set up an Ingress
+controller to route the traffic.
```
Comment (cmoulliard) [inline kind/README.adoc:15]: Remove the Carriage return ;-)

File: kind/README.adoc [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+The goal of this project is to simplify our life when we use kubernetes
+on a laptop. It currently supports the following features: 
+
+* Create/delete a K8s cluster using the latest cluster version, or the one you specify, 
+* Launch a local container registry exposed on port `localhost:5000`, 
+* Set up an Ingress
+controller to route the traffic. 
+
+The bash script has been developed supporting the installation of a docker registry according to the following options:
```
Comment (cmoulliard) [inline kind/README.adoc:17]: `installation of a docker registry according` -> `installation of a container images registry according` to stay consistent with what you said before

File: kind/README.adoc [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+:toc: left
+:toclevels: 2
+
+== Introduction
+
+The goal of this project is to simplify our life when we use kubernetes
+on a laptop. It currently supports the following features: 
+
+* Create/delete a K8s cluster using the latest cluster version, or the one you specify, 
+* Launch a local container registry exposed on port `localhost:5000`,
```
Comment (cmoulliard) [inline kind/README.adoc:13]: `container registry exposed` -> `container images registry exposed`

File: kind/README.adoc [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+* Create/delete a K8s cluster using the latest cluster version, or the one you specify, 
+* Launch a local container registry exposed on port `localhost:5000`, 
+* Set up an Ingress
+controller to route the traffic. 
+
+The bash script has been developed supporting the installation of a docker registry according to the following options: 
+
+* Unsecure: plain HTTP
+* Secured: HTTP/HTTPS endpoints available
```
Comment (cmoulliard) [inline kind/README.adoc:20]: `HTTP/HTTPS endpoints available` -> `HTTP/HTTPS endpoints available and authentication using [htpasswd](https://httpd.apache.org/docs/current/programs/htpasswd.html)`

File: kind/README.adoc [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+== Prerequisite
+
+To be able to install Kind using this project the following tools are required.
+
+* jq: https://stedolan.github.io/jq/download/
+* Kind: https://github.com/kubernetes-sigs/kind/releases
+* Docker desktop: https://www.docker.com/products/docker-desktop
+* openssl
```
Comment (cmoulliard) [inline kind/README.adoc:29]: What about kubectl and helm tools ?

File: kind/README.adoc [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+== Usage
+
+From the repository root execute the following command.
+
+[source,bash]
+----
+./kind/deploy.sh command [parameters,...]
```
Comment (cmoulliard) [inline kind/README.adoc:37]: `./deploy.sh` -> `./kind.sh`

File: kind/README.adoc [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+----
+
+The available commands are:
+
+.Script commands
+[%header,cols="2m,4"]
+|===
+| Command | Meaning
+
+| install | Install the kind cluster
```
Comment (cmoulliard) [inline kind/README.adoc:47]: `Install the kind cluster` -> `Install a kubernetes cluster as container`
`Remove the kind cluster` -> `Remove the kubernetes cluster`

File: kind/README.adoc [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+
+.Script commands
+[%header,cols="2,5"]
+|===
+| Command | Meaning
+
+| `--ingress`
+
+[.fuchsia]#string# / [.red]#required# 
+a| Ingress to be deployed.
```
Comment (cmoulliard) [inline kind/README.adoc:63]: `Ingress to be deployed.` -> `Ingress controller to be deployed.`

File: kind/README.adoc [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+
+* *`docker` <= Default*
+* `podman`
+
+|===
+
+
+== Deploy
+
+Deploying a kind cluster is as simple as executing the following command.
```
Comment (cmoulliard) [inline kind/README.adoc:104]: `Deploying a kind cluster ...` -> `Creating a kubernetes cluster ...`

File: kind/README.adoc [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+[source,bash]
+----
+curl -s -L "https://raw.githubusercontent.com/snowdrop/k8s-infra/main/kind/deploy.sh" | bash -s install --ingress nginx --cluster-name twm --provider podman --port-map "30000:30000,31000:31000"
+----
+
+=== Post deployment steps
+
+Allow using an insecure container registry.
+
+Add the `kind-registry.conf` file to the `/etc/containers/registries.conf.d/` folder having the following contents.
```
Comment (cmoulliard) [inline kind/README.adoc:116]: **Notes**: This is correct for linux users but not for macos users and most probably if you use docker vs podman. To be reviewed please

File: kind/README.adoc [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+[source]
+----
+[[registry]]
+location = "localhost:5000"
+insecure = true
+----
+
+== Undeploy
```
Comment (cmoulliard) [inline kind/README.adoc:125]: As the command to be used is named `remove`, can you rename `undeploy` to `remove` within the document please

File: kind/README.adoc [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+.kind undeploy statement
+[source,bash]
+----
+curl -s -L "https://raw.githubusercontent.com/snowdrop/k8s-infra/main/kind/deploy.sh" | bash -s remove --cluster-name twm
+----
+
+== Secured docker registry
+
+To deploy a local secured (using htpasswd) docker registry use the `--secure-registry` parameter.
```
Comment (cmoulliard) [inline kind/README.adoc:137]: Can we do that independently to install the docker registry as HTTP/HTTPS = TLS registry ?

**Review by cmoulliard — CHANGES_REQUESTED** [review]

**Comment by cmoulliard** [comment]
This text displayed by bash should be reviewed part of the PR `Kind installation script` with something like `Managing a kubernetes cluster using kind ...` @jacobdotcosta

File: kind/deploy.sh (hunk 2 above)
Comment (jacobdotcosta) [reply kind/deploy.sh:140]: The `--kubernetes-version` parameter should be implemented and available for this.

https://github.com/snowdrop/k8s-infra/pull/279/files#diff-e31912bfaa479f5abf54676480bc184cf0cce935c48b30a92a9cb0c48ac982b0R670

```bash
        --kubernetes-version) KUBERNETES_VERSION="$2"; shift ;;
```

Comment (cmoulliard) [reply kind/deploy.sh:140]: LGTM

File: kind/README.adoc (hunk 5 above)
Comment (jacobdotcosta) [reply kind/README.adoc]: done

File: kind/README.adoc (hunk 6 above)
Comment (jacobdotcosta) [reply kind/README.adoc]: done

File: kind/README.adoc (hunk 7 above)
Comment (jacobdotcosta) [reply kind/README.adoc]: done

File: kind/README.adoc (hunk 12 above)
Comment (jacobdotcosta) [reply kind/README.adoc]: done

File: kind/README.adoc (hunk 14 above)
Comment (jacobdotcosta) [reply kind/README.adoc]: done

**Comment by jacobdotcosta** [comment]
> I'm thinking that we should install ingress by default as we need it every time and then you can simplify this command `./deploy.sh install --ingress nginx` @jacobdotcosta

The Ingress installation is mandatory and the `--ingress` parameter is required. The parameter is used to select between `nginx` and `kourier`.

File: kind/README.adoc (hunk 15 above)
Comment (jacobdotcosta) [reply kind/README.adoc:147]: I'm not sure if this is possible as the local registry is included in the kind cluster `yaml` by the `[plugins."io.containerd.grpc.v1.cri".registry.mirrors."localhost:${REGISTRY_PORT}"]
    endpoint = ["http://${registry_name}:${REGISTRY_PORT}"]` directive. I'd have to check if we can install the cluster and then patch the cluster definition by changing this parameter. 

Should I create a new issue to address this or you think not having this is a blocker for rolling out the script as is?

@cmoulliard

Comment (cmoulliard) [reply kind/README.adoc:147]: No.

File: kind/README.adoc (hunk 13 above)
Comment (jacobdotcosta) [reply kind/README.adoc:121]: The [podman installation documentation](https://podman.io/getting-started/installation) doesn't provide information regarding the registry configuration for Mac OS. 

Do you have a `/etc/containers/registries.conf.d/` folder on your machine, @cmoulliard ?

```bash
$ ls -l /etc/containers/
total 8
drwxr-xr-x. 1 root root    0 Feb 14 14:42 certs.d
drwxr-xr-x. 1 root root   14 Feb 14 14:42 oci
-rw-r--r--. 1 root root  569 Feb 14 14:42 policy.json
-rw-r--r--. 1 root root 3897 Feb 14 14:42 registries.conf
drwxr-xr-x. 1 root root   74 Mar 30 12:57 registries.conf.d
drwxr-xr-x. 1 root root  132 Feb 20 15:37 registries.d
drwxr-xr-x  1 root root    0 Feb 14 14:42 systemd
```

Comment (cmoulliard) [reply kind/README.adoc:121]: No as the Macos is not a Linux OS and by consequence do not support natively `containers`. This is why we need the help of docker (able to create locally a VM using Hyperkit).

Certificate generated are stored here:
```
ls -la /etc/docker/certs.d
total 0
drwxr-xr-x  7 root  wheel  224 Mar  8 15:02 .
drwxr-xr-x  3 root  wheel   96 Sep 17  2021 ..
drwxr-xr-x  3 root  wheel   96 Mar  8 15:02 127.0.0.1.sslip.io:5000
drwxr-xr-x  3 root  wheel   96 Mar  8 14:56 127.0.1.sslip.io:5000
drwxr-xr-x  3 root  wheel   96 Sep 17  2021 95.217.159.244:32500
drwxr-xr-x  4 root  wheel  128 Sep 28  2021 kind-registry:5000
drwxr-xr-x  3 root  wheel   96 Jun  2  2022 registry.harbor.192.168.1.90.nip.io
```
and registries here:
```
cat /Users/cmoullia/.docker/config.json
{
  "auths": {
    "ghcr.io": {},
    "https://index.docker.io/v1/": {},
    "index.docker.io": {},
    "localhost:5000": {},
    "quay.io": {},
  },
  "credsStore": "desktop"
}
```
**Remark**: There is also a `$HOME/.docker/certs.d` but I have no idea if it is used !!!

Comment (jacobdotcosta) [reply kind/README.adoc:121]: I've separated the information for insecure registries for docker and podman.

**Pull request merged**